### ISSUE TITLE
authinfo syntax seems to have stopped working sometime in the current emacs development cycle

### ISSUE THREAD

**Issue by emacsomancer** [dengste/org-caldav#320]
Doing things as per the manual, i.e. having an autoinfo with a line like

````
machine www.google.com:443 port https login username password secret
````

works fine for interactionless background syncing in Emacs 29 and even in the current pretest 30; but sometime in the cycle of  the current devel Emacs 31 it does not seem to be able to parse lines with the syntax as above, as it queries everytime for userid and password whenever it tries to sync.

**Comment by dsteuer** [comment]
Stumbled upon the same using a Nextcloud. Furthermore manually logging in no longer seems to work in 31.0.50, but works in 29.4 . The debug buffer only shows "checking connection..." and the password prompt returns ad infinitum.

**Comment by emacsomancer** [comment]
Yes, to clarify, I'm actually having the issue with a Nextcloud as well. (just using the example from the org-caldav manual above.)

**Comment by jackkamm** [comment]
Problem was reported by this user:
https://github.com/dengste/org-caldav/issues/313

I'm still on emacs29. So any help in tracking down the issue is appreciated.

**Comment by real-or-random** [comment]
I'm pretty sure this is the change in https://debbugs.gnu.org/cgi/bugreport.cgi?bug=72526.

Try dropping the port from the host:
```
machine www.google.com port https login username password secret
```

**Comment by dertuxmalwieder** [comment]
This still happens with today's Emacs 31 build, so I guess something's wrong with how org-caldav handles authentication.

My `.authinfo.gpg` contains

    machine CALDAVSERVER port https login USER password PASSWORD

I had briefly thought that this might be related to GnuPG, but a plain-text `.authinfo` does not work either.

**Comment by real-or-random** [comment]
> so I guess something's wrong with how org-caldav handles authentication.
> 
> My `.authinfo.gpg` contains
> 
> ```
> machine CALDAVSERVER port https login USER password PASSWORD
> ```
> 

If `CALDAVSERVER` is like `www.example.com` and not like `www.example.com:443`, then this is unrelated to this issue.

**Comment by dertuxmalwieder** [comment]
It actually is, because ...

> Try dropping the port from the host

... that did **not** fix it for me.

**Comment by real-or-random** [comment]
> > Try dropping the port from the host
> 
> ... that did **not** fix it for me.

This issue is about the need to drop the port from the host. If dropping the port from the host doesn't solve your problem, then I conclude that your problem is unrelated.

**Comment by emacsomancer** [comment]
Problem 1: recent emacs expects the `.authinfo.gpg` file entry to be something like:
`machine dav.mycaldav.org port https login MYLOGIN password MYPASSWD`
(and this is what you should put there, changing `dav.mycaldav.org` to the appropriate server and login & password likewise.)

Problem 2: but org-caldav expects to find an entry where
 the `:host` has the "look-up" form like  `:host dav.mycaldav.org:443`, even
 though the actual value in the data part of the record should be of the
 form `:host dav.mycaldav.org`.

 I.e., the `.authinfo.gpg` line `machine dav.mycaldav.org port https login MYLOGIN password MYPASSWD`
will trigger the generation of an entry like:
````
 (auth-source :host "dav.mycaldav.org" :max 1 :require
                             (:user :secret) :create nil)
                ((:host "dav.mycaldav.org" :port "https" :user "MYUSER"
                       :secret
                        #[0 "\300\242\223\f\0\300\242\302\301!!\202\302\301!\207"
                         [(nil)
                          "uM2sf2fsQqX/mZ4E+eaNdWrZ42YUgC0wNzhasdfadfjafOmM=-JgehaACEZBNjgEsfdfQtoQ=="
                          auth-source--deobfuscate]
                         3])         )
````

but we want something like:
````
 (auth-source :max 1 :host "dav.mycaldav.org:443" :port "https")
                ((:host "dav.mycaldav.org" :port "https" :user "MYUSER"
                        :secret
                        #[0 "\300\242\223\f\0\300\242\302\301!!\202\302\301!\207"
                         [(nil)
                       "uM2sf2fsQqX/mZ4E+eaNdWrZ42YUgC0wNzhasdfadfjafOmM=-JgehaACEZBNjgEsfdfQtoQ=="
                          auth-source--deobfuscate]
                         3])         )
````

 you can inspect the variable `password-data` (`C-h v password-data`) to see this.

# Ok, how to fix it (hacky):

put these somewhere in your config

````elisp
;; for issues with auth-source not resetting
(setq caldav/has-password-cache-been-reset-already nil) ;; set to t once reset
````

````elisp
;; replace "dav.mycaldav.org" string with whatever your dav client address is
(defun caldav/clear-password-cache-put-in-caldav-info ()
  "Clear `password-data' & the auth-source netrc cache, and put in data in format caldav expects."
  (interactive)
  (unless
      (and
       caldav/has-password-cache-been-reset-already
       (not (null
             (password-read-from-cache
              '(auth-source :max 1 :host "dav.mycaldav.org:443" :port "https")))))
    (let* (;; hardcode of key that auth-source-user-and-password creates for dav.mycaldav.org
           (key-for-mycaldav-org
            '(auth-source :host "dav.mycaldav.org" :max 1 :require (:user :secret) :create nil))
           ;; hardcode of key that org-caldav expects dav key to look like
           (key-for-what-org-caldav-expects
            '(auth-source :max 1 :host "dav.mycaldav.org:443" :port "https"))
           ;; temp holder for key for dav.mycaldav.org:
           (tmp-caldav-value nil))
        (progn
          (setq auth-source-netrc-cache nil) ;; clear netrc-cache
          (setq password-data (make-hash-table :test 'equal)) ;; clear out database
          
          ;; trigger password-cache to store dav.mycaldav.org info
          (auth-source-user-and-password "dav.mycaldav.org")
          (setq tmp-caldav-value
                (password-read-from-cache key-for-mycaldav-org))
          (puthash key-for-what-org-caldav-expects tmp-caldav-value password-data)
          ;; indicate that key should be correctly set now and doesn't need resetting
          (setq caldav/has-password-cache-been-reset-already t)))))
````

and then something like this (just adding a hook to the hacky function):

````elisp
;; org-caldav definition
(use-package org-caldav
  :ensure t
  :after (org)
  :init
;; ...  
  (advice-add 'org-caldav-sync :before #'caldav/clear-password-cache-put-in-caldav-info)
  ;; ...
  )
````

**Comment by jackkamm** [comment]
I updated to Emacs 31 and verified that dropping the port from the host made my authinfo work again.

I don't think the advice to modify `password-data` is necessary, authinfo works for me in Emacs 31 without it. And org-caldav does not directly lookup `"dav.mycaldav.org"` or `"dav.mycaldav.org:443"` anywhere (IIRC we rely on `url` package to extract the `:host` key and do the lookup).

I've updated the manual to document the usage of the new authinfo syntax for Emacs 31+.

**Comment by emacsomancer** [comment]
> I updated to Emacs 31 and verified that dropping the port from the host made my authinfo work again.
> 
> I don't think the advice to modify `password-data` is necessary, authinfo works for me in Emacs 31 without it. And org-caldav does not directly lookup `"dav.mycaldav.org"` or `"dav.mycaldav.org:443"` anywhere (IIRC we rely on `url` package to extract the `:host` key and do the lookup).
> 
> I've updated the manual to document the usage of the new authinfo syntax for Emacs 31+.

Maybe the thing with modifying `password-data` is due to some other interaction in my init.el; not sure.
